simple_doc/simple_doc.man
## NAME
  simple_doc - documentation generator for comments in Lua code.

## SYNOPSIS
  `simple_doc [OPTION]... [INPUT-FILE] [OUTPUT-FILE]`

## DESCRIPTION
  Parses docstrings in a Lua source file and writes them to the output file. This is helpful for generating an API reference for functions and fields in some source code. Docstrings are comments that begin with the pattern `---` or `--[[--` without leading characters besides whitespace. The docstring ends at the first non-comment line or the end of the comment block (the pattern `]]`). Docstrings can be basic text that will be copied directly to the output, but annotations can also be used for more advanced behavior.
  
  Annotations are recognized by an "@" symbol at the beginning of a line in the docstring. They are based off of annotations used in [lua-language-server](https://github.com/sumneko/lua-language-server/wiki/Annotations). Only the few listed below are supported and the rest are simply ignored and treated as empty lines.
  
  * `---@docdef [function/variable definition]`
    
    Indicates that a function/variable definition should be inserted at this line. This enables `@param` and `@return` to substitute named parameter/return types into the definition line. If the function/variable definition is not provided, it is guessed from the context (the first non-comment line after the docstring). This annotation can also be used multiple times within a docstring, which is useful for functions with multiple call signatures.
    
  * `---@param <name[?]> <type[|type...]> [description]`
    
    Define a parameter for a function. Putting a question mark (`?`) after the parameter name will mark it as optional, meaning `nil` is an accepted type. If the description is provided it is ignored.
    
  * `---@return <type> [<name> [comment] | [name] #<comment>]`
    
    Define a `return` value for a function. For this one, a question mark after the return type will mark it as optional. If the comment is provided it is ignored.
    
  * `---@see`, `---@author`
    
    Just adds a basic comment that is prefixed with the annotation name.
  
  
  It's easy to embed documentation in an existing file by using the `--insert-start` and `--insert-end` options. For example, this can be used to put Markdown formatted docstrings in a README.md file. Strings like `<!-- SIMPLE-DOC:START (FILE:...) -->` and `<!-- SIMPLE-DOC:END -->` can be placed in the readme to mark where to insert documentation, and the same strings passed to the aforementioned options. This can also be automated (with GitHub actions or other tools) to get the documentation always synced up with the code. Using Markdown docstrings is also nice for editors that can render these on hover, Sublime Text with LSP-lua works well for this purpose.
  
  Some of the concepts and inspiration for this program came from [LDoc](https://github.com/lunarmodules/LDoc).

## OPTIONS
  Input/output files default to `-` to specify standard input/output respectively.
  
  `-h`, `--help`                display help message and exit
  
  `-B`, `--boilerplate=NUMBER`  ignore first N comments in file (boilerplate comments)
  
  `-C`, `--context`             include next line in file after each comment block
  
  `--insert-start=STRING`       insert output text in file starting at the given string
  
  `--insert-end=STRING`         same as insert-start for the ending position
  
  `--ocdoc`                     add bullet points and indent like the OpenComputers docs

## EXAMPLES
  `simple_doc -B --ocdoc test.lua test.out`
  
  Input file `test.lua`:
```lua
--------------------------------------------------------------------------------
-- some boilerplate comments
-- ...
--------------------------------------------------------------------------------

--[[--
Sample
text
with
multiple
lines
]]

--- Sample text 2

local my_module = {}

--- `my_module.message = "hello"`
-- 
-- Configures the message to send.
my_module.message = "hello"

--- `my_module.sendMessage(target: string): boolean`
-- 
-- Sends the message stored in `my_module.message` to a destination specified by
-- `target`. Returns true if the message was sent and false if a transmission
-- error occurred.
function my_module.sendMessage(target)
  print("Sending message ", my_module.message, " to ", target)
  
  -- Do some sending here...
  -- These comments are not docstrings, and don't show up in documentation.
  
  return true
end

---@docdef
-- 
-- Callback that triggers when a message is received from another host. Returns
-- true if message data was processed successfully, or false and an error
-- message otherwise.
-- 
---@param host string
---@param port number
---@param ... string|table
---@return boolean success
---@return string|nil errorMessage
function my_module.onMessageReceived(host, port, ...)
  
  -- Do some stuff...
  
  return true
end
```
  
  
  Output file `test.out`:
```
Sample
text
with
multiple
lines

Sample text 2

* `my_module.message = "hello"`
  
  Configures the message to send.

* `my_module.sendMessage(target: string): boolean`
  
  Sends the message stored in `my_module.message` to a destination specified by
  `target`. Returns true if the message was sent and false if a transmission
  error occurred.

* `my_module.onMessageReceived(host: string, port: number, ...: string|table) -> success: boolean, errorMessage: string|nil`
  
  Callback that triggers when a message is received from another host. Returns
  true if message data was processed successfully, or false and an error
  message otherwise.
```


simple_doc/simple_doc.lua
--------------------------------------------------------------------------------
-- Lua documentation generator.
-- 
-- @see file://simple_doc/README.md
-- @author tdepke2
--------------------------------------------------------------------------------

-- Check for optional dependency unix_compatibility. This is needed only when running simple_doc outside of OpenOS.
do
  pcall(function() require = require("unix_compatibility") end)
end
local fs = require("filesystem")
local shell = require("shell")

local USAGE_STRING = [[Usage: simple_doc [OPTION]... [INPUT-FILE] [OUTPUT-FILE]

Options:
  -h, --help                   display help message and exit
  -B, --boilerplate=NUMBER     ignore first N comments in file (boilerplate comments)
  -C, --context                include next line in file after each comment block
      --insert-start=STRING    insert output text in file starting at the given string
      --insert-end=STRING      same as insert-start for the ending position
      --ocdoc                  add bullet points and indent like the OpenComputers docs

For more information, run: man simple_doc
]]

local args, opts = shell.parse(...)


---@docdef
-- 
-- Wrapper for `io.open()` to check if the file was opened successfully, and
-- raise an error if not.
-- 
---@param filename string
---@param mode? openmode
---@return file*
---@nodiscard
local function ioOpenSafe(filename, mode)
  if fs.isDirectory(filename) then
    io.stderr:write("simple_doc " .. filename .. ": is a directory\n")
    os.exit(2)
  end
  local file, err = io.open(filename, mode)
  if not file then
    io.stderr:write("simple_doc " .. filename .. ": " .. tostring(err) .. "\n")
    os.exit(2)
  end
  return file
end


---@docdef
-- 
-- Searches for annotations in the docSection (patterns beginning with the `@`
-- symbol). Most of these are replaced with empty lines to hide them from
-- documentation, but a few (`@docdef`, `@param`, and `@return`) are used to add
-- function/variable definitions with named parameter types and return values.
-- 
---@param contextLine string
---@param docSection table
local function formatAnnotations(contextLine, docSection)
  -- Tables for tracking function/variable definition lines, named parameters, and potentially named return values.
  local funcs, params, returns
  
  -- Step through each line in docSection to find lines that begin with "@".
  for i = 1, docSection.n do
    local annotation, arguments = string.match(docSection[i], "^%s*@(%S+)%s*(.*)")
    if annotation then
      docSection[i] = ""
      if annotation == "docdef" then
        -- The "@docdef [function/variable definition]" annotation was found. If no definition is provided, guess it from the contextLine.
        if arguments == "" then
          docSection[i] = string.match(contextLine, "function%s+(.*)") or contextLine
          docSection[i] = "`" .. docSection[i] .. "`"
        else
          docSection[i] = arguments
        end
        funcs = funcs or {}
        funcs[#funcs + 1] = i
      elseif annotation == "param" then
        -- The "@param <name[?]> <type[|type...]> [description]" annotation was found. We ignore the description if provided.
        local paramName, paramTypes, paramDesc = string.match(arguments, "(%S+)%s+(%S+)%s*(.*)")
        if paramName then
          params = params or {}
          params[paramName] = paramTypes
        end
      elseif annotation == "return" then
        -- The "@return <type> [<name> [comment] | [name] #<comment>]" annotation was found. We ignore the comment if provided.
        local returnTypes, returnName, returnDesc = string.match(arguments, "(%S+)%s*(%S*)%s*(.*)")
        if returnTypes then
          returns = returns or {}
          returns[#returns + 1] = {returnTypes, returnName}
        end
      elseif annotation == "see" or annotation == "author" then
        docSection[i] = string.upper(string.sub(annotation, 1, 1)) .. string.sub(annotation, 2) .. ": " .. arguments
      end
    end
  end
  
  -- If some function/variable definitions were found, try to add types to parameters and return values for each definition.
  if funcs then
    for _, index in ipairs(funcs) do
      -- Substitute function parameter names with the annotated version where applicable.
      if params then
        local funcParams = string.match(docSection[index], "%(.*%)")
        if funcParams then
          local annotatedParams = ""
          for p in string.gmatch(funcParams, "[%w_%.]+") do
            if params[p] then
              p = p .. ": " .. params[p]
            elseif params[p .. "?"] then
              p = p .. "?: " .. params[p .. "?"]
            end
            annotatedParams = annotatedParams .. p .. ", "
          end
          docSection[index] = string.gsub(docSection[index], funcParams, string.sub(annotatedParams, 1, -3), 1)
        end
      end
      
      -- Add on the list of annotated return values if provided.
      local annotatedReturns = ""
      for _, returnTypeNamePair in ipairs(returns or {}) do
        annotatedReturns = annotatedReturns .. (returnTypeNamePair[2] ~= "" and returnTypeNamePair[2] .. ": " or "") .. returnTypeNamePair[1] .. ", "
      end
      if annotatedReturns ~= "" then
        local trailingChars = string.match(docSection[index], "^.*%)(.+)$")
        if trailingChars then
          docSection[index] = string.sub(docSection[index], 1, -1 - #trailingChars)
        else
          trailingChars = ""
        end
        docSection[index] = docSection[index] .. " -> " .. string.sub(annotatedReturns, 1, -3) .. trailingChars
      end
    end
  end
end


---@docdef
-- 
-- Write a documentation comment block to the output file. Skips any blocks that
-- are considered boilerplate code, adds a newline to separate blocks, and trims
-- trailing empty lines.
-- 
---@param outputFile file*
---@param docSection table
local function writeSection(outputFile, docSection)
  local contextLine = docSection[docSection.n]
  if docSection.n > 0 and not opts["C"] then
    docSection[docSection.n] = nil
    docSection.n = docSection.n - 1
  end
  formatAnnotations(contextLine, docSection)
  
  -- Trim trailing lines that are only whitespace. Return early if docSection is empty.
  for i = docSection.n, 1, -1 do
    if string.find(docSection[i], "%S") then
      break
    end
    docSection[i] = nil
    docSection.n = docSection.n - 1
  end
  if docSection.n == 0 then
    docSection.sectionNumber = docSection.sectionNumber + 1
    return
  end
  
  if opts["B"] and docSection.sectionNumber <= opts["B"] then
    -- Ignore the boilerplate comment.
    for i, v in ipairs(docSection) do
      docSection[i] = nil
    end
  else
    if opts["B"] then
      if docSection.sectionNumber > opts["B"] + 1 then
        outputFile:write("\n")
      end
    elseif docSection.sectionNumber ~= 1 then
      outputFile:write("\n")
    end
    
    if opts["ocdoc"] and docSection[1] and string.find(docSection[1], "^%s*`") then
      -- The ocdoc format adds a bullet point to the first line (if it begins with a backtick) and indents the rest of the block.
      outputFile:write("* ", docSection[1], "\n")
      for i = 2, #docSection do
        outputFile:write("  ", docSection[i], "\n")
        docSection[i] = nil
      end
      docSection[1] = nil
    else
      -- Add remaining lines to output.
      for i, v in ipairs(docSection) do
        outputFile:write(v, "\n")
        docSection[i] = nil
      end
    end
  end
  docSection.sectionNumber = docSection.sectionNumber + 1
  docSection.n = 0
end


---@docdef
-- 
-- Reads the given input file to look for comment blocks formatted as
-- documentation. These are appended to the output file.
-- 
---@param inputFile file*
---@param outputFile file*
local function buildDoc(inputFile, outputFile)
  local state = 0
  local docSection = {sectionNumber = 1, n = 0}
  
  local line, lineNum = inputFile:read(), 1
  while line do
    if state == 1 then
      -- Within comment block (line type).
      local docText = string.match(line, "^%s*%-%-+%s?(.*)")
      if docText then
        docSection.n = docSection.n + 1
        docSection[docSection.n] = docText
      else
        docSection.n = docSection.n + 1
        docSection[docSection.n] = line
        state = 0
        writeSection(outputFile, docSection)
      end
    elseif state == 2 then
      -- Within comment block (multi-line type).
      if not string.find(line, "%]%]") then
        docSection.n = docSection.n + 1
        docSection[docSection.n] = line
      else
        state = 3
      end
    elseif state == 3 then
      -- At line after a multi-line comment block.
      docSection.n = docSection.n + 1
      docSection[docSection.n] = line
      state = 0
      writeSection(outputFile, docSection)
    end
    
    -- Default state, check for comment prefix that could indicate doc comment.
    if state == 0 and string.find(line, "^%s*%-%-") then
      local docText = string.match(line, "^%s*%-%-%-+%s?(.*)")
      if docText then
        state = 1
      else
        docText = string.match(line, "^%s*%-%-%[%[%-%-+%s?(.*)")
        if docText then
          state = 2
        end
      end
      if docText and #docText > 0 then
        docSection.n = docSection.n + 1
        docSection[docSection.n] = docText
      end
    end
    line, lineNum = inputFile:read(), lineNum + 1
  end
  writeSection(outputFile, docSection)
end


---@docdef
-- 
-- Check command line options, open files, generate documentation, and write
-- results to output.
local function main()
  local inputFilename = (args[1] == nil and "-" or args[1])
  local outputFilename = (args[2] == nil and "-" or args[2])
  if #args > 2 or opts["h"] or opts["help"] then
    io.write(USAGE_STRING)
    os.exit(0)
  end
  
  if opts["boilerplate"] then
    opts["B"] = opts["boilerplate"]
  end
  if opts["B"] then
    opts["B"] = math.max(tonumber(opts["B"]) or 1, 0)
  end
  
  if opts["context"] then
    opts["C"] = true
  end
  
  if opts["insert-start"] or opts["insert-end"] then
    if type(opts["insert-start"]) ~= "string" or type(opts["insert-end"]) ~= "string" then
      io.stderr:write("simple_doc: must provide both --insert-start and --insert-end with string values, or skip them both\n")
      os.exit(2)
    end
  end
  
  local inputFile, outputFile
  local outputFileContents
  
  -- Attempt to open input file (or use stdin).
  if inputFilename == "-" then
    inputFile = io.stdin
  else
    inputFilename = shell.resolve(inputFilename)
    inputFile = ioOpenSafe(inputFilename)
  end
  
  -- Attempt to open output file (or use stdout).
  if outputFilename == "-" then
    outputFile = io.stdout
  else
    outputFilename = shell.resolve(outputFilename)
    
    -- If insert-start/insert-end options are provided, read the output file first and look for the start/end strings.
    if opts["insert-start"] then
      outputFile = ioOpenSafe(outputFilename)
      
      local startFound, endFound
      outputFileContents = {n = 0, insertIndex = 0}
      local line = outputFile:read()
      while line do
        if not startFound then
          if line == opts["insert-start"] then
            startFound = true
            outputFileContents.insertIndex = outputFileContents.n + 1
          end
          outputFileContents.n = outputFileContents.n + 1
          outputFileContents[outputFileContents.n] = line
        elseif not endFound then
          if line == opts["insert-end"] then
            endFound = true
            outputFileContents.n = outputFileContents.n + 1
            outputFileContents[outputFileContents.n] = line
          end
        else
          outputFileContents.n = outputFileContents.n + 1
          outputFileContents[outputFileContents.n] = line
        end
        line = outputFile:read()
      end
      
      if not startFound then
        io.stderr:write("simple_doc " .. outputFilename .. ": failed to find insert-start string \"" .. opts["insert-start"] .. "\" in file\n")
        os.exit(2)
      end
      if not endFound then
        io.stderr:write("simple_doc " .. outputFilename .. ": failed to find insert-end string \"" .. opts["insert-end"] .. "\" in file\n")
        os.exit(2)
      end
      outputFile:close()
    end
    outputFile = ioOpenSafe(outputFilename, "w")
  end
  
  if outputFileContents then
    -- Paste contents of outputFile back in until the start string.
    for i = 1, outputFileContents.insertIndex do
      outputFile:write(outputFileContents[i], "\n")
    end
  end
  
  buildDoc(inputFile, outputFile)
  inputFile:close()
  
  if outputFileContents then
    -- Paste remainder of outputFile back in after the end string.
    for i = outputFileContents.insertIndex + 1, outputFileContents.n do
      outputFile:write(outputFileContents[i], "\n")
    end
  end
  outputFile:close()
end

main()
